dali/kernels/imgproc/flip_gpu.cuh

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_FLIP_FLIP_GPU_H
+#define DALI_KERNELS_FLIP_FLIP_GPU_H
+
+#include <cuda_runtime.h>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename T>
+__global__ void FlipKernel(T *__restrict__ output, const T *__restrict__ input, size_t layers,
+                           size_t height, size_t width, size_t channels,
+                           int32 flip_x, int32 flip_y) {
+  size_t xc = blockIdx.x * blockDim.x + threadIdx.x;
+  size_t y = blockIdx.y * blockDim.y + threadIdx.y;
+  size_t z = blockIdx.z * blockDim.z + threadIdx.z;
+  if (xc >= width * channels || y >= height || z >= layers) {
+    return;
+  }
+  size_t channel = xc % channels;
```

Comment:
An idea: for small number of channels you could use a different kernel, with different block layout (but same grid layout) and get rid of division/modulo. This, however, would require having two kernel variants.

Reply:
Idea for a follow-up.